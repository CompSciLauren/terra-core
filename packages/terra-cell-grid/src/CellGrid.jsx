import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './CellGrid.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node.isRequired,
  /**
   * Indicates the desired divider styles to apply to a cell grid and its children.
   * One of `'none'`, `'vertical'`, `'horizontal'`, `'both'`.
   */
  dividerStyle: PropTypes.oneOf(['none', 'vertical', 'horizontal', 'both']),
  /**
  /**
   * Function callback for the ref of the cell grid.
   */
  refCallback: PropTypes.func,
};

const defaultProps = {
  dividerStyle: 'none',
};

const CellGrid = ({
  children,
  dividerStyle,
  refCallback,
  ...customProps
}) => {
  const cellGridClassNames = cx([
    `divider-${dividerStyle}`,
    'cell-grid',
    customProps.className,
  ]);

  return (
    <div {...customProps} className={cellGridClassNames} ref={refCallback}>
      {children}
    </div>
  );
};

CellGrid.propTypes = propTypes;
CellGrid.defaultProps = defaultProps;

export default CellGrid;
