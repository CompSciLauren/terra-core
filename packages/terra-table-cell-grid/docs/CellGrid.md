# Cell Grid

The CellGrid is a sub component that groups table cells passed through children. Each row handles it's own selection state through props. The onSelect function that can be passed to the cell grid will return the metaData associated to the cell grid. The metaData allows for greater control of state with the ability to associate any indicator or function with the cell grid.

## Usage

```jsx
import React from 'react';
import { CellGrid, Cell } from 'terra-table-cell-grid';

  <CellGrid
    key="unique-react-key"
    isSelectable
    isSelected
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Cell>
      Cell Content
    </Cell>
    <Cell>
      Cell Content
    </Cell>
  </CellGrid>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
