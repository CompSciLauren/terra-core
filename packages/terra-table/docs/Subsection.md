# Subsection

The Terra Table Subsection is a sub component that creates a subsection header based on the given props and groups the child cell grids beneath it. The subsection header has the ability to toggle the display of its child rows given the isCollapsible and isCollapsed props. For a subsection to be collapsed both isCollapsible and isCollapsed must be true. Each section handles its collapsible and collapsed state.

## Usage

```jsx
import React from 'react';
import { CellGrid, Subsection } from 'terra-table';

  <Subsection
    key="unique-react-key"
    title="Subsection Title"
    isCollapsible
    isCollapsed
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Row />
    <Row />
    <Row />
  </Subsection>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
