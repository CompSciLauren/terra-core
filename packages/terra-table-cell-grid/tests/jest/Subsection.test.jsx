import React from 'react';
import Subsection from '../../src/Subsection';
import CellGrid from '../../src/CellGrid';

// Snapshot test
it('should render a default subsection', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with a cellgrid', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test">
      <CellGrid />
    </Subsection>,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with isCollpased', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" isCollapsed />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with isCollapsible', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" isCollapsible />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with level', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" level={4} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with metaData', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" metaData={{}} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with onSelect', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" onSelect={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a subsection with refCallback', () => {
  const section = shallow(
    <Subsection colSpan={1} title="test" refCallback={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});
