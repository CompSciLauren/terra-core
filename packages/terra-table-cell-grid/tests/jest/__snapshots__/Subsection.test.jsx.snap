// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default subsection 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with a cellgrid 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
  <CellGrid
    dividerStyle="none"
    isSelectable={false}
    isSelected={false}
    isStriped={false}
    selectionStyle="default"
  />
</Fragment>
`;

exports[`should render a subsection with isCollapsible 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={true}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with isCollpased 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={true}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with level 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={4}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with metaData 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    metaData={Object {}}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with onSelect 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    onSelect={[MockFunction]}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with refCallback 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    refCallback={[MockFunction]}
    title="test"
  />
</Fragment>
`;
